### ISSUE TITLE
Runtime server terminates when client does not follow protocol

### ISSUE THREAD

**Issue by twisteroidambassador** [mobvoi/wenet#322]
**Describe the bug**
When a WebSocket client does not follow the protocol exactly, the entire runtime server terminates after printing this message:

    terminate called without an active exception

**To Reproduce**
1. Prepare a WebSocket client that deviates from the protocol. For example:
    * Not sending the {"signal": "end"} message before closing WebSocket (This also applies to a compliant client that closes connection too early, connections lost due to network conditions, etc.)
    * Sending unrecognized TEXT messages such as non-JSON messages, empty JSON messages, etc.
2. Connect to a runtime server with such a client
3. Watch the runtime server terminate

**Expected behavior**
The runtime server should not terminate because of a misbehaving client.

**Additional context**
The cause for this particular problem is that `ConnectionHandler::decode_thread_` is neither joined nor detached when the `ConnectionHandler` instance is being destructed. See https://en.cppreference.com/w/cpp/thread/thread/%7Ethread

The code in `operator()` only reaches `decode_thread_->join()` if the request ends normally, either by `"signal": "end"` or by receiving another BINARY message after endpointing. In other cases, such as when the client closes the WebSocket connection, or when `OnError` closes the WebSocket connection, `operator()` skips the cleanup.

`ConnectionHandler` should join `decode_thread_` in all code paths.

**Comment by robin1001** [comment]
Yes, we haven't deal the boundary conditions and error handling now.
Will try to refine it in the future.

**Comment by yuekaizhang** [comment]
> **Describe the bug**
> When a WebSocket client does not follow the protocol exactly, the entire runtime server terminates after printing this message:
> 
> ```
> terminate called without an active exception
> ```
> 
> **To Reproduce**
> Steps to reproduce the behavior:
> 
> 1. Prepare a WebSocket client that deviates from the protocol. For example:
>    
>    * Not sending the {"signal": "end"} message before closing WebSocket (This also applies to a compliant client that closes connection too early, connections lost due to network conditions, etc.)
>    * Sending unrecognized TEXT messages such as non-JSON messages, empty JSON messages, etc.
> 2. Connect to a runtime server with such a client
> 3. Watch the runtime server terminate
> 
> **Expected behavior**
> The runtime server should not terminate because of a misbehaving client.
> 
> **Additional context**
> The cause for this particular problem is that `ConnectionHandler::decode_thread_` is neither joined nor detached when the `ConnectionHandler` instance is being destructed. See https://en.cppreference.com/w/cpp/thread/thread/%7Ethread
> 
> The code in `operator()` only reaches `decode_thread_->join()` if the request ends normally, either by `"signal": "end"` or by receiving another BINARY message after endpointing. In other cases, such as when the client closes the WebSocket connection, or when `OnError` closes the WebSocket connection, `operator()` skips the cleanup.
> 
> `ConnectionHandler` should join `decode_thread_` in all code paths.

@twisteroidambassador Any follow-ups? I was wondering if you fix this issue.

**Comment by AdolfVonKleist** [comment]
Modifying this catch should help:
* https://github.com/mobvoi/wenet/blob/main/runtime/core/websocket/websocket_server.cc#L238

join there and don't forget whatever is left.  This was what I did with an older version to resolve unceremonious closures from the client end that were killing the wss server.  It would probably need more review now as it predates the recent updates for continuous decoding, etc.:
```catch (beast::system_error const& se) {
    // This indicates that the session was closed
    if (decode_thread_ != nullptr) {
      decode_thread_->join();
    }
    if (se.code() != websocket::error::closed) {
      OnSpeechEnd();
      LOG(ERROR) << se.code().message();
    }
  }
```

**Comment by Mddct** [comment]
Because detach is used, when the child thread is abnormal, such as ws is closed, it is still writing, but there is no join.  Then the parent site will crash

**Comment by Johnsonys** [comment]
Thanks @AdolfVonKleist for the suggestion! Besides, we find there is one more step needs to be done. When executing
`decode_thread_->join();` the decoder thread will try to write the final result back to client page while the ws is closed, so we just simply skip the ws.write step.
